src/integrity-checker.js

```diff
@@ -63,51 +63,42 @@ export default class InstallationIntegrityChecker {
 
   config: Config;
 
-  async _getModuleLocation(usedRegistries?: Set<RegistryNames>): Promise<string> {
-    // build up possible folders
-    let registries = registryNames;
-    if (usedRegistries && usedRegistries.size > 0) {
-      registries = usedRegistries;
-    }
-    const possibleFolders = [];
-    if (this.config.modulesFolder) {
-      possibleFolders.push(this.config.modulesFolder);
-    }
-
-    // ensure we only write to a registry folder that was used
-    for (const name of registries) {
-      const loc = path.join(this.config.cwd, this.config.registries[name].folder);
-      possibleFolders.push(loc);
-    }
+  /**
+   * Get the common ancestor of every node_modules - it might be a node_modules directory itself, but isn't required to.
+   */
 
-    // if we already have an integrity hash in one of these folders then use it's location otherwise use the
-    // first folder
-    let loc;
-    for (const possibleLoc of possibleFolders) {
-      if (await fs.exists(path.join(possibleLoc, constants.INTEGRITY_FILENAME))) {
-        loc = possibleLoc;
-        break;
-      }
+  _getModulesRootFolder(): string {
+    if (this.config.modulesFolder) {
+      return this.config.modulesFolder;
+    } else if (this.config.workspaceRootFolder) {
+      return this.config.workspaceRootFolder;
+    } else {
+      return path.join(this.config.lockfileFolder, 'node_modules');
     }
-
-    return loc || possibleFolders[0];
   }
 
   /**
-   * Get the location of an existing integrity hash. If none exists then return the location where we should
-   * write a new one.
+   * Get the directory in which the yarn-integrity file should be written.
    */
 
-  async _getIntegrityHashLocation(usedRegistries?: Set<RegistryNames>): Promise<IntegrityHashLocation> {
-    let locationFolder;
-
-    if (this.config.enableMetaFolder) {
-      locationFolder = path.join(this.config.lockfileFolder, constants.META_FOLDER);
+  _getIntegrityFileFolder(): string {
+    if (this.config.modulesFolder) {
+      return this.config.modulesFolder;
+    } else if (this.config.enableMetaFolder) {
+      return path.join(this.config.lockfileFolder, constants.META_FOLDER);
     } else {
-      locationFolder = await this._getModuleLocation(usedRegistries);
+      return path.join(this.config.lockfileFolder, 'node_modules');
```

Comment:
No, workspaceRootFolder should only be used for resolving relative paths. In workspaces it will be the package root folder, and we don't want to store the `.yarn-integrity` file there.